client/src/main/scala/org/http4s/netty/client/Http4sHandler.scala

```diff
@@ -21,41 +21,109 @@ import cats.effect.Async
-import io.netty.channel.ChannelHandlerContext
-import io.netty.channel.ChannelInboundHandlerAdapter
+import io.netty.channel._
 import io.netty.handler.codec.http.HttpResponse
 import io.netty.handler.timeout.IdleStateEvent
-import org.http4s.Response
+import org.http4s._
+import org.http4s.client.RequestKey
 import org.http4s.netty.NettyModelConversion
 
 import java.io.IOException
+import java.nio.channels.ClosedChannelException
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.util.Failure
+import scala.util.Success
 
-private[netty] class Http4sHandler[F[_]](cb: Http4sHandler.CB[F], dispatcher: Dispatcher[F])(
-    implicit F: Async[F])
+private[netty] class Http4sHandler[F[_]](dispatcher: Dispatcher[F])(implicit F: Async[F])
     extends ChannelInboundHandlerAdapter {
 
-  private[this] val logger = org.log4s.getLogger
-  val modelConversion = new NettyModelConversion[F]
+  private val logger = org.log4s.getLogger
+  private val modelConversion = new NettyModelConversion[F]
+  private val promises =
+    collection.mutable.Queue[Either[Throwable, Resource[F, Response[F]]] => Unit]()
+  // By using the Netty event loop assigned to this channel we get two benefits:
+  //  1. We can avoid the necessary hopping around of threads since Netty pipelines will
+  //     only pass events up and down from within the event loop to which it is assigned.
+  //     That means calls to ctx.read(), and ct.write(..), would have to be trampolined otherwise.
+  //  2. We get serialization of execution: the EventLoop is a serial execution queue so
+  //     we can rest easy knowing that no two events will be executed in parallel.
+  private var eventLoopContext: ExecutionContext = _
+  private var ctx: ChannelHandlerContext = _
+
+  private var pending: Future[Unit] = Future.unit
+
+  private[netty] def dispatch(request: Request[F]): Resource[F, Response[F]] = {
+    val key = RequestKey.fromRequest(request)
+
+    modelConversion
+      .toNettyRequest(request)
+      .evalMap { nettyRequest =>
+        F.async[Resource[F, Response[F]]] { cb =>
+          addCallback(cb).addListener { (future: ChannelFuture) =>
+            void {
+              if (future.isCancelled || !future.isSuccess) {
+                ctx.fireChannelInactive()
+              } else {
+                logger.trace(s"Sending request to $key")
+                ctx.writeAndFlush(nettyRequest)
+                logger.trace(s"After request to $key")
+              }
+            }
+          }
+          F.delay(Some(F.delay(ctx.close()).liftToF))
+        }
+      }
+      .flatMap(identity)
+  }
+
+  private def addCallback(
+      callback: Either[Throwable, Resource[F, Response[F]]] => Unit): ChannelFuture =
+    if (ctx.executor().inEventLoop()) {
+      promises.enqueue(callback)
+      ctx.newSucceededFuture()
+    } else {
+      ctx.newPromise().addListener((_: ChannelFuture) => promises.enqueue(callback)).setSuccess()
+    }
```

Comment:
I think there are some race conditions we need to guard against such as making sure that once we enter the netty executor the pipeline is still alive. Otherwise, that callback is going to go into a dead queue and never be handled.

It might help to consolidate all of that noise into a `def safeDispatch` method that takes the `cb`, the request to send, etc. In the `F.async` block above you can simply check if you're in the executor and if so call `safeDispatch` right away, and otherwise do `ctx.executor.submit(() => safeDispatch(cb, nettyRequest))`. You're effectively already doing the latter via the `ChannelPromise` stuff you're doing, this is just more direct.

Reply:
good feedback. have added a commit that adresses that